### DISCUSSION TITLE
Update pillow to import as pillow and follow PEP-8 naming conventions

### DISCUSSION THREAD

**Discussion by sodul** [python-pillow/Pillow#8060] [Ideas]
The pillow package is a fork of PIL from 2010. While PIL did not follow the pep-8 naming conventions I feel that with the improved linting tools in the python ecosystem it might be time for pillow to mature from its original roots and start adopting pep-8.

I think this could be done in a 100% backward compatible manner and just publish pillow so that both `from PIL import Image` (confusing) and `from pillow import image` (pep-8 compliant) will work. A new pillow release could add the new compliant names and over a few releases mark the old names as deprecated, and a few years later completely remove support of the old, non compliant names.

I know this suggestion might be controversial, but folks that do not know that `PIL` is really `pillow` do get confused. If you are not previously familiar with PIL and look for a `PIL` library on pypi.org, there are dozens of matches and it is not very clear that you want `pillow`. Indirectly this is a security concern as it facilitates a supply chain attack when someone would want to add `PIL` to their project and innocently install a compromised version of PIL that would look like a better match than `pillow`.

https://pypi.org/search/?q=pil  ... pillow is not even on the first 10 pages.

**Comment by Yay295** [comment]
> `from pillow import image` (pep-8 compliant)

Except it's not. `image` should be `Image`. What issues other than the module name do you know of?

edit: I feel like I should also mention the very first thing in PEP 8 after the introduction: https://peps.python.org/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds

**Comment by sodul** [comment]
This is ultimately the project maintainers decision whether to standardize on PEP 8 or not for the existing codebase. I do understand that it would take some time to refactor the style and provide backward compatibility. I just wanted to share that the naming convention is unusual compared to the vast majority of python projects and that linting tools such as pylint, ruff, pep8 will need to be tuned accordingly for folks that would contribute.

An other example in `Image.py` is this method:
```python
def isImageType(t: Any) -> TypeGuard[Image]:
```

Which under PEP 8 guidance would probably be written as:
```python
def _is_image_type(obj: Any) -> TypeGuard[Image]:
```

This way linters would warn external callers that they are not supposed to use it.

Again, just a suggestion to help make PIL/pillow more accessible to new users. Maybe I should have opened this under Discussions.

**Comment by aclark4life** [comment]
@sodul Thank you, converted to discussion! It's interesting to discuss but probably not actionable for the core team at the moment. Also, this is at least two separate tasks: lower casing Image module and renaming PIL package to pillow … neither of which we've ever seriously considered.

**Reply by sodul** [reply]
For context the reason I thought about starting this conversation is that I was not familiar with PIL. I've worked with python since py2.3 but I rarely deal with images, so when a team member asked me to add PIL to our requirements I did not know that I needed to look for `pillow` since it is not listed in the first pages when looking for `PIL`. Maybe the PIL page on pypi.org could be updated to mention and link to the pillow project as a very simple short term improvement.

The idea to rename the modules is that internally we follow the Google style guide, which does not allow to import classes and methods, only modules and packages. So when I saw the `from PIL import Image` this confused me as well.

I do understand the desire to stay backward compatible, this is very useful, but I also actually enjoy that Python and the community no longer try to stay backward compatible with old ways forever. The new releases of Python are more aggressive about deprecating older packages and older ways of doing things. Tools such as black are helping make the python community more consistent and I think this is an improvement.

I do not have much skin in the game, I've shared my input on this.

**Reply by radarhere** [reply]
> Maybe the PIL page on pypi.org could be updated to mention and link to the pillow project as a very simple short term improvement.

Apparently the only way to update the description is to [release a new version](https://github.com/python-pillow/Pillow/issues/1535#issuecomment-820397804).

**Reply by sodul** [reply]
This could be done with a -post1 release which is specifically intended to update metadata and not code changes.

From your link it seems that one concern is that `pip install PIL` should keep on breaking, which can easily be done by having an empty package that would require python 1.0 or some silly dependency that would guarantee installation breakage and also not install any actual files.

https://peps.python.org/pep-0440/#post-releases
